// @metaplex-foundation/mpl-core-candy-machine/dist/src/generated/programs/mplCoreCandyMachineCore.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMplCoreCandyMachineCoreProgramId = exports.getMplCoreCandyMachineCoreProgram = exports.createMplCoreCandyMachineCoreProgram = exports.MPL_CORE_CANDY_MACHINE_CORE_PROGRAM_ID = void 0;
const errors_1 = require("../errors");
exports.MPL_CORE_CANDY_MACHINE_CORE_PROGRAM_ID = 'CMACYFENjoBMHzapRXyo1JZkVS6EtaDDzkjMrmQLvr4J';
function createMplCoreCandyMachineCoreProgram() {
    return {
        name: 'mplCoreCandyMachineCore',
        publicKey: exports.MPL_CORE_CANDY_MACHINE_CORE_PROGRAM_ID,
        getErrorFromCode(code, cause) {
            return (0, errors_1.getMplCoreCandyMachineCoreErrorFromCode)(code, this, cause);
        },
        getErrorFromName(name, cause) {
            return (0, errors_1.getMplCoreCandyMachineCoreErrorFromName)(name, this, cause);
        },
        isOnCluster() {
            return true;
        },
    };
}
exports.createMplCoreCandyMachineCoreProgram = createMplCoreCandyMachineCoreProgram;
function getMplCoreCandyMachineCoreProgram(context, clusterFilter) {
    return context.programs.get('mplCoreCandyMachineCore', clusterFilter);
}
exports.getMplCoreCandyMachineCoreProgram = getMplCoreCandyMachineCoreProgram;
function getMplCoreCandyMachineCoreProgramId(context, clusterFilter) {
    return context.programs.getPublicKey('mplCoreCandyMachineCore', exports.MPL_CORE_CANDY_MACHINE_CORE_PROGRAM_ID, clusterFilter);
}
exports.getMplCoreCandyMachineCoreProgramId = getMplCoreCandyMachineCoreProgramId;
