// @metaplex-foundation/mpl-core-candy-machine/dist/src/mintV1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mintV1 = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const mintV1_1 = require("./generated/instructions/mintV1");
const guards_1 = require("./guards");
const hooked_1 = require("./hooked");
function mintV1(context, input) {
    const { mintArgs = {}, group = (0, umi_1.none)(), ...rest } = input;
    // Parsing mint data.
    const program = context.programs.get('mplCoreCandyGuard');
    const candyMachine = (0, umi_1.publicKey)(input.candyMachine, false);
    const mintContext = {
        minter: input.minter ?? context.identity,
        payer: input.payer ?? context.payer,
        asset: (0, umi_1.publicKey)(input.asset, false),
        candyMachine,
        candyGuard: (0, umi_1.publicKey)(input.candyGuard ?? (0, hooked_1.findCandyGuardPda)(context, { base: candyMachine }), false),
    };
    const { data, remainingAccounts } = (0, guards_1.parseMintArgs)(context, program, mintContext, mintArgs);
    const ix = (0, mintV1_1.mintV1)(context, {
        ...rest,
        mintArgs: data,
        group,
    }).items[0];
    const [keys, signers] = (0, guards_1.parseGuardRemainingAccounts)(remainingAccounts);
    ix.instruction.keys.push(...keys);
    ix.signers.push(...signers);
    return (0, umi_1.transactionBuilder)([ix]);
}
exports.mintV1 = mintV1;
