// @metaplex-foundation/mpl-core-candy-machine/dist/src/defaultGuards/assetPaymentMulti.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.assetPaymentMultiGuardManifest = void 0;
const generated_1 = require("../generated");
const guards_1 = require("../guards");
/**
 * The assetPayment guard restricts the mint to holders of a predefined
 * Collection and burns the holder's asset(s) when minting.
 *
 */
exports.assetPaymentMultiGuardManifest = {
    name: 'assetPaymentMulti',
    serializer: generated_1.getAssetPaymentMultiSerializer,
    mintParser: (context, mintContext, args) => {
        const remainingAccounts = [];
        remainingAccounts.push({
            publicKey: args.requiredCollection,
            isWritable: true,
        });
        remainingAccounts.push({ publicKey: args.destination, isWritable: true });
        args.assets.forEach((asset) => {
            remainingAccounts.push({ publicKey: asset, isWritable: true });
        });
        return { data: new Uint8Array(), remainingAccounts };
    },
    routeParser: guards_1.noopParser,
};
