// @metaplex-foundation/mpl-core-candy-machine/dist/src/generated/instructions/initializeCandyMachine.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.initializeCandyMachine = exports.getInitializeCandyMachineInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getInitializeCandyMachineInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 8 })],
        ['itemsAvailable', (0, serializers_1.u64)()],
        ['maxEditionSupply', (0, serializers_1.u64)()],
        ['isMutable', (0, serializers_1.bool)()],
        ['configLineSettings', (0, serializers_1.option)((0, types_1.getConfigLineSettingsSerializer)())],
        ['hiddenSettings', (0, serializers_1.option)((0, types_1.getHiddenSettingsSerializer)())],
    ], { description: 'InitializeCandyMachineInstructionData' }), (value) => ({
        ...value,
        discriminator: [175, 175, 109, 31, 13, 152, 155, 237],
        maxEditionSupply: value.maxEditionSupply ?? 0,
        isMutable: value.isMutable ?? true,
        configLineSettings: value.configLineSettings ?? (0, umi_1.none)(),
        hiddenSettings: value.hiddenSettings ?? (0, umi_1.none)(),
    }));
}
exports.getInitializeCandyMachineInstructionDataSerializer = getInitializeCandyMachineInstructionDataSerializer;
// Instruction.
function initializeCandyMachine(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplCoreCandyMachineCore', 'CMACYFENjoBMHzapRXyo1JZkVS6EtaDDzkjMrmQLvr4J');
    // Accounts.
    const resolvedAccounts = {
        candyMachine: {
            index: 0,
            isWritable: true,
            value: input.candyMachine ?? null,
        },
        authorityPda: {
            index: 1,
            isWritable: true,
            value: input.authorityPda ?? null,
        },
        authority: { index: 2, isWritable: false, value: input.authority ?? null },
        payer: { index: 3, isWritable: true, value: input.payer ?? null },
        collection: { index: 4, isWritable: true, value: input.collection ?? null },
        collectionUpdateAuthority: {
            index: 5,
            isWritable: true,
            value: input.collectionUpdateAuthority ?? null,
        },
        mplCoreProgram: {
            index: 6,
            isWritable: false,
            value: input.mplCoreProgram ?? null,
        },
        systemProgram: {
            index: 7,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        sysvarInstructions: {
            index: 8,
            isWritable: false,
            value: input.sysvarInstructions ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.authorityPda.value) {
        resolvedAccounts.authorityPda.value = (0, hooked_1.findCandyMachineAuthorityPda)(context, { candyMachine: (0, shared_1.expectPublicKey)(resolvedAccounts.candyMachine.value) });
    }
    if (!resolvedAccounts.authority.value) {
        resolvedAccounts.authority.value = context.identity.publicKey;
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.mplCoreProgram.value) {
        resolvedAccounts.mplCoreProgram.value = context.programs.getPublicKey('mplCore', 'CoREENxT6tW1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d');
        resolvedAccounts.mplCoreProgram.isWritable = false;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    if (!resolvedAccounts.sysvarInstructions.value) {
        resolvedAccounts.sysvarInstructions.value = (0, umi_1.publicKey)('Sysvar1nstructions1111111111111111111111111');
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getInitializeCandyMachineInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.initializeCandyMachine = initializeCandyMachine;
