// @metaplex-foundation/mpl-core-candy-machine/dist/src/generated/instructions/mintAssetFromCandyMachine.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.mintAssetFromCandyMachine = exports.getMintAssetFromCandyMachineInstructionDataSerializer = void 0;
const mpl_core_1 = require("@metaplex-foundation/mpl-core");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const shared_1 = require("../shared");
function getMintAssetFromCandyMachineInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 8 })],
        ['plugins', (0, serializers_1.array)((0, mpl_core_1.getPluginAuthorityPairSerializer)())],
    ], { description: 'MintAssetFromCandyMachineInstructionData' }), (value) => ({
        ...value,
        discriminator: [84, 175, 211, 156, 56, 250, 104, 118],
    }));
}
exports.getMintAssetFromCandyMachineInstructionDataSerializer = getMintAssetFromCandyMachineInstructionDataSerializer;
// Instruction.
function mintAssetFromCandyMachine(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplCoreCandyMachineCore', 'CMACYFENjoBMHzapRXyo1JZkVS6EtaDDzkjMrmQLvr4J');
    // Accounts.
    const resolvedAccounts = {
        candyMachine: {
            index: 0,
            isWritable: true,
            value: input.candyMachine ?? null,
        },
        authorityPda: {
            index: 1,
            isWritable: true,
            value: input.authorityPda ?? null,
        },
        mintAuthority: {
            index: 2,
            isWritable: false,
            value: input.mintAuthority ?? null,
        },
        payer: { index: 3, isWritable: true, value: input.payer ?? null },
        assetOwner: {
            index: 4,
            isWritable: false,
            value: input.assetOwner ?? null,
        },
        asset: { index: 5, isWritable: true, value: input.asset ?? null },
        collection: { index: 6, isWritable: true, value: input.collection ?? null },
        mplCoreProgram: {
            index: 7,
            isWritable: false,
            value: input.mplCoreProgram ?? null,
        },
        systemProgram: {
            index: 8,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        sysvarInstructions: {
            index: 9,
            isWritable: false,
            value: input.sysvarInstructions ?? null,
        },
        recentSlothashes: {
            index: 10,
            isWritable: false,
            value: input.recentSlothashes ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.authorityPda.value) {
        resolvedAccounts.authorityPda.value = (0, hooked_1.findCandyMachineAuthorityPda)(context, { candyMachine: (0, shared_1.expectPublicKey)(resolvedAccounts.candyMachine.value) });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.mplCoreProgram.value) {
        resolvedAccounts.mplCoreProgram.value = context.programs.getPublicKey('mplCore', 'CoREENxT6tW1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d');
        resolvedAccounts.mplCoreProgram.isWritable = false;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    if (!resolvedAccounts.sysvarInstructions.value) {
        resolvedAccounts.sysvarInstructions.value = (0, umi_1.publicKey)('Sysvar1nstructions1111111111111111111111111');
    }
    if (!resolvedAccounts.recentSlothashes.value) {
        resolvedAccounts.recentSlothashes.value = (0, umi_1.publicKey)('SysvarS1otHashes111111111111111111111111111');
    }
    if (!resolvedArgs.plugins) {
        resolvedArgs.plugins = [];
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getMintAssetFromCandyMachineInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.mintAssetFromCandyMachine = mintAssetFromCandyMachine;
